src/rust/engine/serverset/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+        backoff_ratio
+      ));
+    }
+    let backoff_ratio =
+      num_rational::Ratio::<i8>::approximate_float(backoff_ratio).ok_or_else(|| {
+        format!(
+          "Couldn't find reasonable backoff ratio for {}",
+          backoff_ratio
+        )
+      })?;
+    let backoff_ratio =
+      num_rational::Ratio::new(*backoff_ratio.numer() as u32, *backoff_ratio.denom() as u32);
+    Ok(Serverset {
+      inner: Arc::new(Inner {
+        servers: servers
+          .into_iter()
+          .map(|s| Backend {
+            server: s,
+            unhealthy_info: Arc::new(Mutex::new(None)),
+          }).collect(),
+        next: AtomicUsize::new(0),
+        failure_initial_lame: initial_lame,
+        failure_backoff_ratio: backoff_ratio,
+        failure_max_lame: max_lame,
+      }),
+    })
+  }
+
+  ///
+  /// Get the next (probably) healthy backend to use.
+  ///
+  /// The caller will be given a backend to use, and should call the supplied callback with the
+  /// observed health of that backend.
+  ///
+  /// If the callback is not called, the health status of the server will not be changed from its
+  /// last known status.
+  ///
+  /// If all resources are unhealthy, this function will block the calling thread until the backoff
+  /// period has completed. We'd probably prefer to use some Future-based scheduling, but that
+  /// would require this type to be Resettable because of our fork model, which would be very
+  /// complex.
+  ///
+  pub fn next(&self) -> (T, Box<Fn(Health) + Send + Sync>) {
+    let (i, server) = loop {
+      let i = self.inner.next.fetch_add(1, Ordering::Relaxed) % self.inner.servers.len();
+      let server = &self.inner.servers[i];
+      let unhealthy_info = server.unhealthy_info.lock();
+      if let Some(ref unhealthy_info) = *unhealthy_info {
+        if unhealthy_info.unhealthy_since.elapsed() < unhealthy_info.next_attempt_after {
+          continue;
+        }
+      }
+      break (i, server);
+    };
+
+    let serverset: Serverset<T> = (*self).clone();
+
+    let callback = Box::new(move |health: Health| serverset.callback(i, health));
+
+    (server.server.clone(), callback)
+  }
+
+  fn callback(&self, server_index: usize, health: Health) {
+    let mut unhealthy_info = self.inner.servers[server_index].unhealthy_info.lock();
+    match health {
+      Health::Unhealthy => {
+        if unhealthy_info.is_some() {
+          if let Some(ref mut unhealthy_info) = *unhealthy_info {
+            unhealthy_info.unhealthy_since = Instant::now();
+            // failure_backoff_ratio's numer and denom both fit in u8s, so hopefully this won't
+            // overflow of lose too much precision...
+            unhealthy_info.next_attempt_after *= *self.inner.failure_backoff_ratio.numer();
+            unhealthy_info.next_attempt_after /= *self.inner.failure_backoff_ratio.denom();
+            unhealthy_info.next_attempt_after = std::cmp::min(
```

Comment:
This series of assignments seems odd to me, why not calculate the correct value and then assign it once? As far as I can tell we won't have any write race conditions because we lock unhealty_info, we won't have read RCs because `self.inner` is atomic, and we won't have read-after-write hazards on `self.inner.failure_backoff_ratio` because it doesn't change except for initialization. And even if it did change, the correct behavior would presumably to make a thread-local copy first to read from.

Reply:
Done